proposals/0540-vision-for-layout-conformance-parity.md

```diff
@@ -0,0 +1,238 @@
+
+## Choosing the capabilities to build
+
+React Native does not implement the full set of browser CSS features. Apart from higher level concepts like stylesheets, selectors, and cascading (which can largely be supplanted by CSS-in-JS solutions shared between RN and browser), React Native doesn’t implement all style properties.
+
+It’s clear there is a desire for more capabilities, but there is a question of which should be added. React Native should enable skills to be transferable, offer a stellar developer experience, and enable substantial code-sharing (while being leaner than a browser). A natural conflict is whether to prioritize building for compatibility (existing usages), or for developer experience. [Adoption of conformant layout](#restriction-to-conformant-trees) is dependent on building out features developers would like to use, in their existing React Native projects. 
+
+With the initial focus of building for developer experience, instead of compatibility, we need input on what the community cares about. We can derive a lot of this from asking directly, but should be opportunistic in our choices, accounting for implementation complexity.
+
+## Mitigating bloat to heap usage
+
+Yoga internally represents a stylesheet as a packed structure of all possible properties. This structure is already 192 bytes, scaling with each additional Yoga Node. As we add more styles, this could expand. Doubling the number of styles on a tree of 20,000 nodes would increase memory consumption by almost 4MB, a huge amount for some of the devices React Native targets. We should measure, and potentially switch YGStyle to a data structure which is pay-for-play in number of properties. This shift would be performance sensitive, but theoretically quick data structures exist (e.g. a bitfield header with a small-vector for larger values). 
+
+## Example capabilities
+
+The following is a non-exhaustive list of capabilities where Yoga lags browsers. We should consult specs used by yoga (CSS Box Alignment, CSS Box Sizing, CSS Flexbox, etc) for a more exhaustive list.
```

Comment:
Isn't this only relevant if you support float? What stops us doing independent subtree layout already?

Reply:
It's relevant for content sizing (e.g. `auto` flex-basis). If a node has `contain: layout` on it then the layout algorithm can assume that the content size of that node is zero. I believe this effectively turns multi-pass layout into single-pass layout for the affected node.

Reply:
My understanding is `contain: layout` requires you to explicitly set `width` and `height`. But if you've done that, and as long as you don't invalidate those values with stuff like `{min,max}{Width,Height}`, and your `flex{,Shrink,Grow}` values are all zero, I think you can apply the same optimisations as if `contain: layout` was set: you should be able to compute the whole subtree independently to the parent (i.e. another thread).

Because we're a subset of CSS and don't support `float`, I don't think we actually get any more guarantees with `contain: layout`

Reply:
Ok, so it turns out that `contain: layout` does not have the effect I was expecting, but `contain: size` does.

Here's an example where `contain: size` differs from `width: 0px` for CSS grid:
```html
<div style="display: grid; grid-template-columns: auto;grid-template-rows: 40px 40px">
  <div style="contain: size">
    <div style="width: 200px"></div>
  </div>
  <div style="width: 100px"></div>
</div>
```

With `contain: size` you get a width of `100px` for first child. With `width: 0px` you get a width of `0px` (and with neither you get a width of `200px`). Conceptually this represents the notion "I should be sized by the container, but my content size should not be taken into account when doing this sizing"

And here's an example with Flexbox:

```html
<div style="display: flex; flex-direction: column; align-items: stretch; height: 100px">
  <div style="contain: size;flex: 1">
    <div style="width: 200px"></div>
  </div>
  <div style="width: 100px; flex: 1"></div>
</div>
```
(here it affect cross-axis size)

This is perhaps a bit more niche than I thought (some if not cases can indeed be achieved by application of already existing styles), but I think it is worth considering.

Reply:
@nicoburns Am I missing from the examples? https://jsfiddle.net/0qbs8pku/2/ - if I take out `contain: size`, nothing changes

Reply:
@jacobp100 Ah, try adding `position: absolute` to the root node of each example (or some other attribute which prevents the root node from stretch-fitting into it's parent. Apologies: my test env had a stylesheet that was adding this for me.

Reply:
Ah yeah that works - can definitely see the difference now. Thanks for the example!